Getting Started Documentation (#118)
Azure/benchpress: docs/getting_started.md

```PowerShell
BeforeAll {
  Import-Module "../BenchPress/Helpers/Azure/BenchPress.Azure.psd1"
}

Describe 'Verify Container Registry Exists' {
  it 'Should contain a container registry with the given name' {
    #arrange
    $rgName = "rg-test"
    $acrName = "acrbenchpresstest1"

    #act
    $exists = Get-AzBPContainerRegistry -ResourceGroupName $rgName -Name $acrName

    #assert
    $exists | Should -Not -BeNullOrEmpty
  }
}

Describe 'Verify Container Registry Exists' {
  it 'Should contain a container registry with the given name' {
    #arrange
    $rgName = "rg-test"
    $acrName = "acrbenchpresstest1"

    #act
    $exists = Get-AzBPContainerRegistryExist -ResourceGroupName $rgName -Name $acrName

    #assert
    $exists | Should -Be $true
  }
}

Describe 'Spin up , Tear down Container Registry' {
  it 'Should deploy a bicep file.' {
    #arrange
    $resourceGroupName = "rg-test"
    $bicepPath = "./containerRegistry.bicep"
    $params = @{
      name           = "acrbenchpresstest2"
      location       = "westus3"
    }

    #act

>     $deployment = Deploy-AzBPBicepFeature -BicepPath $bicepPath -Params $params -ResourceGroupName $resourceGroupName

Reviewer: The deployment and setup would probably be better demonstrated in a BeforeAll block to be consistent with best practices for organizing Pester Tests.

Reviewer: Ah, yeah, fair point. I was thinking more along the lines of this being *only* documentation. But, I forgot that the source of the documentation is the original files. I'm OK with leaving it the way it is so that it is consistent with the files.

    #assert
    $deployment.ProvisioningState | Should -Be "Succeeded"

    #clean up

>     Remove-AzBPBicepFeature -ResourceGroupName $resourceGroupName

Reviewer: Similar to the above, the removal would probably best be demonstrated in an AfterAll block.

    $acrName = "acrbenchpresstest1"

    #act
    $exists = Get-AzBPContainerRegistryExist -ResourceGroupName $rgName -Name $acrName

    #assert
    $exists | Should -Be $true
  }
}
```

This test looks very similar to the first test, except it is using the `Get-AzBPContainerRegistryExist` helper instead. This helper doesn't return the object representing the Container Registry, but instead returns true if the Container Registry exists and false if it does not exists. Assuming the container registry exists, we assert that the return should be true.

Now let's look at the third `Describe` block:

```PowerShell
Describe 'Spin up , Tear down Container Registry' {
  it 'Should deploy a bicep file.' {
    #arrange
    $resourceGroupName = "rg-test"
    $bicepPath = "./containerRegistry.bicep"
    $params = @{
      name           = "acrbenchpresstest2"
      location       = "westus3"
    }

    #act
    $deployment = Deploy-AzBPBicepFeature -BicepPath $bicepPath -Params $params -ResourceGroupName $resourceGroupName

    #assert
    $deployment.ProvisioningState | Should -Be "Succeeded"

    #clean up
    Remove-AzBPBicepFeature -ResourceGroupName $resourceGroupName
  }
}
```

> This test is unlike the first two tests because we are using the `Deploy-AzBPBicepFeature` helper to deploy a bicep file. `Deploy-AzBicepFeature` will take a bicep file and deploy it to an existing resource group, and it will return the deployment details. As you can see, we instead assert that the deployment should be successful. We then tear down the deployed resources using the `Remove-AzBPBicepFeature` helper.

Reviewer: If the recommendations above are followed, then this will change.

## Fixing the Tests

Now that we've done a walkthrough of the three tests, let's fix them.

> The first two tests assumed that our container container registry was already deployed to a resource group. However, we never deployed the `containerRegistry.bicep` file ourselves! The third test assumed we had an existing resource group to deploy to, but we never deployed that either! Let's go ahead and fix these assumptions now:

Reviewer: Grammar/Spelling: `container container`

1. Create a resource group in your subscription.

> 1. Deploy the container registry bicep file to that resource group: `az deployment group create --resource-group <your-resource-group-name> --template-file .\containerRegistry.bicep` (You could also use the `Deploy-AzBicepFeature` helper here)

Reviewer: Instead of using the `az` CLI, should this use the Azure PowerShell Module to be consistent?

Reviewer: Also, do we need to do this step if the Pester test already spins up the resources?

Author: > Instead of using the `az` CLI, should this use the Azure PowerShell Module to be consistent?

Great idea! Changed over to New-AzResourceGroupDeployment cmdlet.

> Also, do we need to do this step if the Pester test already spins up the resources?

Resolved in other comments.

# Getting Started with BenchPress

## BenchPress intro

The basic usage pattern for writing automated tests with BenchPress is to:

1. Write tests using [Pester testing framework](https://pester.dev/docs/quick-start). Create test files ending in
   `Tests.ps1` and run test files using `Invoke-Pester` .

> 1. Deploy resources to Azure using bicep (or helpers from BenchPress)

Reviewer 2: NIT:
```suggestion
1. Deploy resources to Azure using bicep (or helpers from BenchPress).
```

1. Use BenchPress to return information about deployed resources.
1. (Optional) Use BenchPress to tear down resources at the end of the test.

## Requirements

> BenchPress uses PowerShell and the Azure Az PowerShell module. It is also expected of users of BenchPress to use Pester

Reviewer 2: ```suggestion
BenchPress uses PowerShell and the Azure Az PowerShell module. Users should use Pester
```
